src/megaapi_impl.cpp

```diff
@@ -25184,14 +25304,73 @@ void MegaFolderDownloadController::cancel()
     //remove subtransfers from pending transferQueue
     megaApi->cancelPendingTransfersByFolderTag(tag);
 
+    //remove ongoing subtransfers
+    long long cancelledSubTransfers = 0;
+    DBTableTransactionCommitter committer(client->tctable);
     while (!subTransfers.empty())
     {
         auto subTransfer = *subTransfers.begin();
-        subTransfer->setState(MegaTransfer::STATE_COMPLETED);
-        DBTableTransactionCommitter committer(client->tctable);
-        megaApi->fireOnTransferFinish(subTransfer, make_unique<MegaErrorPrivate>(API_EINCOMPLETE), committer);
+
+        Transfer *transfer = subTransfer->getTransfer();
+        if (!transfer)
+        {
+            LOG_warn << "Subtransfer without attached Transfer for folder transfer: " << subTransfer->getFileName();
+
+            subTransfer->setState(MegaTransfer::STATE_COMPLETED); //TODO: copmleted?
+            megaApi->fireOnTransferFinish(subTransfer, make_unique<MegaErrorPrivate>(API_EINCOMPLETE), committer);
+
+            continue;
+        }
+
+#ifdef _WIN32
+        if (transfer->type == GET)
+        {
+            transfer->localfilename.append("", 1);
+            WIN32_FILE_ATTRIBUTE_DATA fad;
+            if (GetFileAttributesExW((LPCWSTR)transfer->localfilename.data(), GetFileExInfoStandard, &fad))
+                SetFileAttributesW((LPCWSTR)transfer->localfilename.data(), fad.dwFileAttributes & ~FILE_ATTRIBUTE_HIDDEN);
+            transfer->localfilename.resize(transfer->localfilename.size()-1);
+        }
+#endif
+
+        MegaErrorPrivate megaError(API_EINCOMPLETE);
+        subTransfer->setLastError(&megaError);
+
+        bool found = false;
+        file_list files = transfer->files;
+        file_list::iterator iterator = files.begin();
+        while (iterator != files.end())
+        {
+            File *file = *iterator;
+            iterator++;
+            if (file->tag == subTransfer->getTag())
+            {
+                found = true;
+                if (!file->syncxfer)
+                {
+                    client->stopxfer(file, nullptr); //No commiter, we are here within sendPendingRequest
```

Comment:
Not sure if the comment is correct here. Why should the committer be invalid here? Shouldn't we use the committer for the cancellation of all transfers declared at the beginning of this method?
(same here: https://github.com/meganz/sdk/pull/2087/files#diff-c3de3d92d9d9472cc3d715cb978d2940R23937)

Reply:
The committer parameters are really just to remind us that one should be on the stack before we make the call - I see we have created one outside the loop anyway,  there's no harm in passing &committer here, that would be the normal way to use it, so it all looks normal.   There's no harm in having more than one on the stack either.  I would have made the parameter a reference rather than pointer if we didn't have the case of these being called from destructors (with committers higher on the stack already)  thanks

Reply:
mmm the problem with having a commiter is that it turns into an assert failing:
stopxfer -> removeTransferFile -> filecachedel -> checkCommiter:

```
void DbTable::checkCommitter(DBTableTransactionCommitter* committer)
{
    assert(!committer || committer == mTransactionCommitter);
}``` 
that is why i passed nullptr (since it is not required)

Reply:
I've added a new commit to have a workaround. resusing the table commiter if one was present

Reply:
Ah I see, you've reminded me - the assert is there to alert us when we ended up with two committers on the stack (not that it's really a problem, but just to keep things tidy).  And it's a case where we've been called from a destructor (~MegaTransferPrivate(), which calls cancel() on the upload/download controller), which is why the first committer on the stack hasn't been passed in here.